WASM: Add SHA* hash functions

akoeplinger opened the pull request:
This ports the managed SHA* hash functions from https://github.com/microsoft/referencesource so we can use them on Browser/WASM where we don't have OpenSSL.

Enables the System.Security.Cryptography.Algorithms tests for these classes: `Tests run: 271, Errors: 0, Failures: 0, Skipped: 1. Time: 11.857137s`

Resolves https://github.com/dotnet/runtime/issues/40076

(related: implement this using the web crypto APIs in .NET6: https://github.com/dotnet/runtime/issues/40074)

EgorBo commented on src/mono/mono/utils/mono-md5.c:
```diff
@@ -109,11 +103,6 @@ mono_md5_init (MonoMD5Context *ctx)
-		ctx->doByteReverse = 1;		
-	else 
-		ctx->doByteReverse = 0;	
 }
```
Just wondering: why did you drop BE support? 🙂 I know it's not supported officially but what about the community support?

EgorBo replied:
Ah I see, you didn't

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/tests/IncrementalHashTests.cs:
```diff
@@ -154,6 +159,7 @@ public static void VerifyEmptyHash(HashAlgorithm referenceAlgorithm, HashAlgorit
         }
 
         [Theory]
+    [   SkipOnMono("Not supported on Browser", TestPlatforms.Browser)]
```
Nit: 

```suggestion
        [SkipOnMono("Not supported on Browser", TestPlatforms.Browser)]
```

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/tests/IncrementalHashTests.cs:
```diff
@@ -219,6 +226,7 @@ public static void AppendDataAfterHashClose()
         }
 
         [Fact]
+       [SkipOnMono("Not supported on Browser", TestPlatforms.Browser)]
```
Nit: 

```suggestion
        [SkipOnMono("Not supported on Browser", TestPlatforms.Browser)]
```

blowdart commented:
Absolutely **NOT**

At least not until someone ran this past legal, for crypto export regulations.

blowdart commented:
Seeing as you added me

I question the need for any of this. Is there a driving customer requirement for parity in WASM?

GrabYourPitchforks commented:
Marking this _no merge_ for now while we work through compliance-related details.

bfmsoft commented:
> Seeing as you added me
> 
> I question the need for any of this. Is there a driving customer requirement for parity in WASM?

Yes, this is needed

vargaz commented on src/mono/mono/metadata/icall-def-netcore.h:
```diff
@@ -1,3 +1,10 @@
+#if defined(TARGET_WASM)
```
Can't we use a managed md5 implementation as well ?

akoeplinger replied:
we could but there's no managed implementation in referencesource so we'd have to write/find one. since this is still being discussed I didn't want to spend time on it, just demonstrate that it works.

vargaz replied:
We have an old mono implementation in mcs/class/corlib/System.Security.Cryptography/MD5CryptoServiceProvider.cs
which has been in use for 15 years.

GrabYourPitchforks commented:
Unmarking _no merge_ per email thread from this morning.

akoeplinger committed:
WASM: Add SHA* hash functions

This ports the managed SHA* hash functions from https://github.com/microsoft/referencesource so we can use them on Browser/WASM where we don't have OpenSSL.

Enables the System.Security.Cryptography.Algorithms tests for these classes: `Tests run: 271, Errors: 0, Failures: 0, Skipped: 1. Time: 11.028473s`

Resolves https://github.com/dotnet/runtime/issues/40076

akoeplinger commented:
Updated the PR:

- removed MD5, SubtleCrypto doesn't expose MD5 so we wouldn't be able to reimplement it on top of SubtleCrypto in net6
- added buffering for the input data and call the hashing function in one go, this mimics what we'll have to do with SubtleCrypto in net6

akoeplinger commented:
/backport to release/5.0-rc2
